
# Intermediate HW
Change the treshold from 0.5 to 0.2, and rerun the code.

Give a commentary in plain English about how that changed precision and recall. What does that mean? What is now included that wasn't before? What part of it is good? What is bad from our Task perspective. Remember: our task was to identify Dissatisfied reviews.

https://en.wikipedia.org/wiki/Precision_and_recall

# HW

In [46]:
##Change the treshold from 0.5 to 0.2, and rerun the code.
pred2A_train = prob2_train > 0.2

def analyze_training_example_2A(i):
    print("Training example number", i)
    print("True rating = %f stars" % Y_train[i])
    print("Expected to be dissatisfied:", pred2A_train[i])
    print("Expected probability of being dissatisfied : %f" % prob2_train[i])
    print("Features = %f / %f" % (X_train[i,0], X_train[i,1]))
    print("Review text = %s" % apps_train[i]['reviewText'])
    
analyze_training_example_2A(max_prob2)
analyze_training_example_2A(min_prob2)

Training example number 6520
True rating = 2.000000 stars
Expected to be dissatisfied: True
Expected probability of being dissatisfied : 0.996439
Features = 0.000000 / 1.000000
Review text = boring
Training example number 853
True rating = 5.000000 stars
Expected to be dissatisfied: False
Expected probability of being dissatisfied : 0.000193
Features = 1.000000 / 0.000000
Review text = Love it!


### HW text
Precision reduced and recall increased! Whenever there is more than 20% probability that the review might be bad, we are flagging it, 20% is a small number honestly to flag a review. On the otherhand, the more we are trying to help people and manually check reviews(even with 80% good probability to be a good review), the more time consuming human work has to be done. So the power of machine learning and automation isn't used to it's potential. It might be good that we are certainly getting most bad reviews and checking them, so that there are no loopholes, but from task perspective the bad part is the reduced automation of flagging.

In [56]:
from sklearn.metrics import precision_score, recall_score
precision2 = precision_score(D_train, pred2_train)
recall2 = recall_score(D_train, pred2_train)
print("For the default threshold (0.5) we get precision = %f "
      "and recall = %f" % (precision2, recall2))

For the default threshold (0.5) we get precision = 0.650405 and recall = 0.227147


<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/2/26/Precisionrecall.svg/330px-Precisionrecall.svg.png"/>

#### We only found 22% of the dissatisfied customers, missed 78% of dissatisfied customers(false negatives).
#### Out of the ones we flagged, 65% are actually dissatified, 35% are actually okay.

## HW precision recall check (verify with text)

In [55]:
from sklearn.metrics import precision_score, recall_score
precision2A = precision_score(D_train, pred2A_train)
recall2A = recall_score(D_train, pred2A_train)
print("For the default threshold (0.2) we get precision = %f "
      "and recall = %f" % (precision2A, recall2A))

For the default threshold (0.2) we get precision = 0.371991 and recall = 0.841968


## Precision REDUCED from 65% to 37% 
## Recall INCREASED from 22% to 84%